server/service/client_mdm.go

```diff
@@ -30,3 +43,148 @@ func (c *Client) RequestAppleCSR(email, org string) (*fleet.AppleCSR, error) {
 	err := c.authenticatedRequest(request, verb, path, &responseBody)
 	return responseBody.AppleCSR, err
 }
+
+func (c *Client) GetBootstrapPackageMetadata(teamID uint) (*fleet.MDMAppleBootstrapPackage, error) {
+	verb, path := "GET", fmt.Sprintf("/api/latest/fleet/mdm/apple/bootstrap/%d/metadata", teamID)
+	request := bootstrapPackageMetadataRequest{}
+	var responseBody bootstrapPackageMetadataResponse
+	err := c.authenticatedRequest(request, verb, path, &responseBody)
+	return responseBody.Metadata, err
+}
+
+func (c *Client) DeleteBootstrapPackage(teamID uint) error {
+	verb, path := "DELETE", fmt.Sprintf("/api/latest/fleet/mdm/apple/bootstrap/%d", teamID)
+	request := deleteBootstrapPackageRequest{}
+	var responseBody deleteBootstrapPackageResponse
+	err := c.authenticatedRequest(request, verb, path, &responseBody)
+	return err
+}
+
+func (c *Client) UploadBootstrapPackage(pkg *fleet.MDMAppleBootstrapPackage) error {
+	verb, path := "POST", "/api/latest/fleet/mdm/apple/bootstrap"
+
+	var b bytes.Buffer
+	w := multipart.NewWriter(&b)
+
+	// add the package field
+	fw, err := w.CreateFormFile("package", pkg.Name)
+	if err != nil {
+		return err
+	}
+	if _, err := io.Copy(fw, bytes.NewBuffer(pkg.Bytes)); err != nil {
+		return err
+	}
+
+	// add the team_id field
+	if err := w.WriteField("team_id", fmt.Sprint(pkg.TeamID)); err != nil {
+		return err
+	}
+
+	w.Close()
+
+	response, err := c.doContextWithBodyAndHeaders(context.Background(), verb, path, "",
+		b.Bytes(),
+		map[string]string{
+			"Content-Type":  w.FormDataContentType(),
+			"Accept":        "application/json",
+			"Authorization": fmt.Sprintf("Bearer %s", c.token),
+		},
+	)
+	if err != nil {
+		return fmt.Errorf("do multipart request: %w", err)
+	}
+
+	var bpResponse uploadBootstrapPackageResponse
+	if err := c.parseResponse(verb, path, response, &bpResponse); err != nil {
+		return fmt.Errorf("parse response: %w", err)
+	}
+
+	return nil
+}
+
+func (c *Client) EnsureBootstrapPackage(bp *fleet.MDMAppleBootstrapPackage, teamID uint) error {
+	isFirstTime := false
+	oldMeta, err := c.GetBootstrapPackageMetadata(teamID)
+	if err != nil {
+		// not found is OK, it means this is our first time uploading a package
+		if !errors.Is(err, notFoundErr{}) {
+			return fmt.Errorf("getting bootstrap package metadata: %w", err)
+		}
+		isFirstTime = true
+	}
+
+	if !isFirstTime {
+		// compare checksums, if they're equal then we can skip the package upload.
+		if bytes.Equal(oldMeta.Sha256, bp.Sha256) {
+			return nil
+		}
```

Comment:
@noahtalerman @lukeheath a small thing that I think solves our concern with CI build times, this will upload the bootstrap package only if the contents of the `.pkg` file have changed since the last time it was uploaded.